streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2133,6 +2167,21 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
       } yield pull
     }
 
+  /**
+   * Fails with the error `None` if value is `Left`.
+   */
+  final def right[O1, O2](implicit ev: O <:< Either[O1, O2]): ZStream[R, Option[E], O2] =
+    self.mapError(Some(_)).rightOrFail(None)
+
+  /**
+   * Fails with fiven error 'e' if value is `Left`.
+   */
+  final def rightOrFail[O1, O2, E1 >: E](e: => E1)(implicit ev: O <:< Either[O1, O2]): ZStream[R, E1, O2] =
+    self.flatMap(ev(_) match {
+      case Right(value) => ZStream.succeed(value)
+      case Left(_)      => ZStream.fail(e)
+    })
```

Comment:
Would it be possible to use `mapM` here? It'd be vastly more efficient as:
```suggestion
    self.mapM(ev(_) match {
      case Right(value) => ZIO.succeed(value)
      case Left(_)      => ZIO.fail(e)
    })
```

Reply:
Let me take a look at that